tree/tree/src/InternalTreeUtils.cxx

```diff
@@ -0,0 +1,194 @@
+         }
+         const auto *realName = chainFiles->First()->GetName();
+         friendNames.emplace_back(std::make_pair(realName, alias));
+         // Get filenames stored in the title member
+         for (const auto *f : *chainFiles) {
+            fileNames.emplace_back(f->GetTitle());
+         }
+      } else {
+         // Get name of the tree
+         const auto realName = GetTreeFullPaths(*frTree)[0];
+         friendNames.emplace_back(std::make_pair(realName, alias));
+
+         // Get filename
+         const auto *f = frTree->GetCurrentFile();
+         if (!f)
+            throw std::runtime_error("A TTree in the list of friends is not linked to any file. "
+                                     "Friends with no associated files are not supported.");
+         fileNames.emplace_back(f->GetName());
+      }
+   }
+
+   return RFriendInfo{std::move(friendNames), std::move(friendFileNames)};
+}
+
+////////////////////////////////////////////////////////////////////////////////
+/// \fn std::vector<std::string> GetTreeFullPaths(const TTree &tree)
+/// \ingroup tree
+/// \brief Retrieve the full path(s) to a TTree or the trees in a TChain.
+/// \param[in] tree The tree or chain from which the paths will be retrieved.
+/// \throws std::runtime_error If the input tree is a TChain but no files could
+///         be found associated with it.
+/// \return If the input argument is a TChain, returns a vector of strings with
+///         the name of the tree of each file in the chain. If the input
+///         argument is a TTree, returns a vector with a single element that is
+///         the full path of the tree in the file (e.g. the name of the tree
+///         itself or the path with the directories inside the file). Finally,
+///         the function returns a vector with just the name of the tree if it
+///         couldn't do any better.
+std::vector<std::string> GetTreeFullPaths(const TTree &tree)
+{
+   // Case 1: this is a TChain. For each file it contains, GetName returns the name of the tree in that file
+   if (auto chain = dynamic_cast<const TChain *>(&tree)) {
+      const auto *chainFiles = chain->GetListOfFiles();
+      if (!chainFiles || chainFiles->GetEntries() == 0) {
+         throw std::runtime_error("The input TChain does not contain any file.");
+      }
+      std::vector<std::string> treeNames;
+      for (const auto *f : *chainFiles)
+         treeNames.emplace_back(f->GetName());
+
+      return treeNames;
+   }
+
+   // Case 2: this is a TTree: we get the full path of it
+   if (const auto *motherDir = tree.GetDirectory()) {
+      // We have 2 subcases (ROOT-9948):
+      // - 1. motherDir is a TFile: return the name of the tree.
+      // - 2. motherDir is a directory: reconstruct the path to the tree in the directory.
+      // Use dynamic_cast to check whether the directory is a TFile
+      if (dynamic_cast<const TFile *>(motherDir)) {
```

Comment:
Another alternative :)
```suggestion
      if (motherDir->GetMotherDir() == nullptr) {
```

Reply:
I'm not very used to the internals of TDirectory, but I see `GetMotherDir` is defined as
```
TDirectory *GetMotherDir() const { return !fMother ? nullptr : dynamic_cast<TDirectory*>(fMother); }
```
So it's still another dynamic_cast, plus makes the intent of this line of code a bit less clear imho. Still nice to know there is this alternative :+1: